ttnn/cpp/ttnn/operations/experimental/ccl/ccl_op_fusion.hpp

```diff
@@ -0,0 +1,117 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "tt_metal/host_api.hpp"
+#include "tt_metal/impl/program/program.hpp"
+
+namespace ttnn {
+namespace experimental {
+namespace ccl {
+
+struct AllGatherFusedOpSignaler {
+
+    std::vector<CoreCoord> fused_op_receiver_cores;
+    std::vector<CoreCoord> fused_op_receiver_cores_noc;
+    std::vector<uint32_t> fused_op_receiver_signal_semaphores;
+
+    /* All Gather specific */
+    std::vector<CoreCoord> all_gather_worker_cores_noc;
+    uint32_t all_gather_worker_sync_semaphore;
+
+    uint32_t initialized = 0;
+
+    AllGatherFusedOpSignaler() {}
```

Comment:
correct me if I'm wrong but we should be able to delete the empty constructor here if we remove the assignment I pointed out in `all_gather_op_multi_core.cpp`. If this is absolutely needed, we should initialize `initialized` in the constructor and then also initialize to 1 in the constructor below (no need to increment at that point).

Also there is a usability bug here. I'd much rather just have an init for each type of initialization needed and check all were triggered rather than a counter which aliases the different types (if you call the same one a couple times you'll think you're initialized, when you aren't.

A preferable way IMO would be to have two separate signalers (emitter/receiver), and make the matmul/consumer op produce the `signal_receiver` and make the allgather/producer op producer the `signal_producer` and then you could construct a higher level object out of the both of them if needed to then get rid of this initialization logic.

I think though we should definitely cleanup how `initialized` is initialized and used (my first sets of comments). My second set of comments is a bit more forward looking and I'd opt not to do that for now as it's more dependent on the final TTNN op fusion design.

Reply:
fixed the bug with initialization. Now there's a separate init check for each init method.